### ISSUE TITLE
Model as the single source of truth

### ISSUE THREAD

**Issue by davaya** [spdx/spdx-3-model#412]
> Please keep in mind that the human-editable markdown files were not meant to be used for anything else besides the spec-parser. That program will produce lots of generated stuff: for example ontologies, but also complete markdown files.  
I'm now seeing more and more attempts to parse these files, which I think it's a step to the wrong direction.

_Originally posted by @zvr in https://github.com/spdx/spdx-3-model/issues/407#issuecomment-1624229910_

There are three steps to using the human-editable markdown files:
1) parsing
2) validation
3) processing (producing lots of generated stuff)

The spec-parser should be simple, its task is to read the markdown files into program variables so that the model can be validated and then processed.  Or the spec parser can include both parsing and validation, so the model used by processing applications is already checked for correctness.

The spec-parser should not do the processing step, that's the job of processing applications that use it.

So "more and more attempts to parse these files" is a step in the right direction if they 1) come up with correct output, and 2) produce better results.  The serialization-playground has a [spec-parser](https://github.com/davaya/spdx-3-serialization-prototype-playground/blob/code/model/parse_model.py) that does parsing - it's a simple task that can be easily checked for correctness by examining the [parsed model](https://github.com/davaya/spdx-3-serialization-prototype-playground/blob/code/model/generated_20230705T100948Z/modelTypes.json).

Is the simple spec-parser correct?  Is there anything in the entire model that it does not parse correctly?  Can processing applications read the output of the simple parser and have every single bit of information that is contained in the model in order to produce everything including "ontologies, and also complete markdown files" that they need to produce (if the Summary and Description text were included instead of being discarded for brevity)?

This is what the simple parser parsed from [Relationship.md](https://github.com/spdx/spdx-3-model/blob/main/model/Core/Classes/Relationship.md):
```
  "Relationship": {
    "Summary": {},
    "Description": {},
    "Metadata": {
      "name": "Relationship",
      "SubclassOf": "Element",
      "Instantiability": "Concrete",
      "_modelRef": "https://rdf.spdx.org/v3/Core/Relationship",
      "_profile": "Core",
      "_category": "Classes",
      "_file": "Relationship.md",
      "_html": "https://github.com/spdx/spdx-3-model/blob/main/model/Core/Classes/Relationship.md"
    },
    "Properties": {
      "from": {
        "type": "Element",
        "minCount": "1",
        "maxCount": "1"
      },
      "to": {
        "type": "Element",
        "minCount": "0"
      },
      "relationshipType": {
        "type": "RelationshipType",
        "minCount": "1",
        "maxCount": "1"
      },
      "completeness": {
        "type": "RelationshipCompleteness",
        "minCount": "0",
        "maxCount": "1"
      },
      "startTime": {
        "type": "DateTime",
        "minCount": "0",
        "maxCount": "1"
      },
      "endTime": {
        "type": "DateTime",
        "minCount": "0",
        "maxCount": "1"
      }
    }
  },
```
Is there anything incorrect?  Is there anything missing?  If not, then the simple parser does step 1 correctly.

Two data points indicate that the simple approach is a step in the right direction:
1) https://github.com/spdx/spdx-3-model/pull/395 - the simple parser found a simple cut-and-paste error that had gone unnoticed when using the spec-parser repo.  An argument can be made that simple code is both easier to check for correctness and easier to understand the functionality it provides.  But regardless of that argument, the simple parser found a simple error and flagged it to be fixed; the code in the spec-parser repo did not.
2) The Relationship class above contains an error.  That error exists in the model files, it was parsed correctly by the simple parser, and when viewed by looking at "[stuff generated by processing applications](https://github.com/davaya/spdx-3-serialization-prototype-playground/blob/code/model/generated_20230705T100948Z/Core/Relationship.md)", the error is more noticeable, and was in fact noticed by @maxhbr.  The generated stuff was generated from the simple parser, and it contains information (links to both the model files and the model version) that is not available from the spec-parser repo code.  That error had gone unnoticed for as long as the model files and the spec-parser repo have been used, but was immediately noticed by Max.

Who else sees the error?

## [Relationship](https://github.com/spdx/spdx-3-model/blob/main/model/Core/Classes/Relationship.md)
Model: [828388b](https://github.com/spdx/spdx-3-model/commit/828388b98c2374f1af6b760ab87fee0d4a11e3f4) 2023-07-05T10:09:48Z
```
class Relationship:
    from: SpdxId = None                                # * 
    to: SpdxId = None                                  # * optional Set[1..*]
    relationshipType: RelationshipType = None          # 
    completeness: RelationshipCompleteness = None      # optional 
    startTime: DateTime = None                         # optional 
    endTime: DateTime = None                           # optional 
    spdxId: SpdxId = None                              # * 
    name: String = None                                # * optional 
    summary: String = None                             # * optional 
    description: String = None                         # * optional 
    comment: String = None                             # * optional 
    creationInfo: CreationInfo = None                  # 
    verifiedUsing: IntegrityMethod = None              # optional Set[1..*]
    externalReference: ExternalReference = None        # optional Set[1..*]
    externalIdentifier: ExternalIdentifier = None      # optional Set[1..*]
    extension: Extension = None                        # optional Set[1..*]
    namespaces: NamespaceMap = None                    # optional Set[1..*]
    imports: ExternalMap = None                        # optional Set[1..*]
```

So I disagree that separating the parsing, validating, and processing tasks is a step in the wrong direction.  The parser should parse.  It's got one job, and it should do it well, simply and understandably.  Ontology generation, documentation generation, etc. use the output of a parser, but they are processing applications, not parsers.

**Comment by goneall** [comment]
One consideration is how standardized we want to source markdowns to be.

When we originally discussed using the input markdown approach, I was thinking it would translate quickly into a different well documented and well maintained "standard" which would then be used by downstream "processors".

The advantage to this approach is several fold:
- It allows us to change the input markdown format to meet the needs of human writability (one of the original goals)
- It allows downstream processors to depend on a well documented standard which will not change frequently
- Downstream processors can take advantage of existing parsers and validation libraries

We've chose OWL/SHACL as the intermediate standard.  I've just started writing a processor which takes the OWL/SHACL input and produces Java source code.  I've previously written OWL/SHACL to JSON validation.  In a previous post, I suggested we could use the input markdown files for JSON schemas which led to some of the additional code spec-parser.  I'm now wondering if it would have been better to use the OWL/SHACL output.  Similarly, for generating documentation.

In summary, I think we should split this discussion into 2 parts:
- Should we "standardize" the input markdown format so it can be used for multiple processors or should we translate to a standard format and use that for downstream processors?
- If we keep the decision to use a standard format for downstream processors, do we keep the decision to use OWL/SHACL?

@davaya - Have you looked at using the OWL/SHACL output as inputs into your serialization data schemas?

**Comment by zvr** [comment]
The **model** is the single source of truth -- and if it's not correct, it should be fixed.

The model is _produced_ by `spec_parser`, a custom software reading a collection of files in an ad-hoc (and unspecified) syntax.

The model is expressed primarily in RDF (with OWL and SHACL).
Since we did not want to impose that everyone should learn these, the ad-hoc syntax was created as a convenience. The only reason for this ad-hoc syntax to exist is the spec-parser and there is no need for anything else to process these files.

The model is also expressed in other formats, also produced by spec_parser: a collection of md files, a collection of html files, a PDF file, a Word document, a JSON schema, a graphical representation, etc. etc. Each of these serves a different purpose (e.g., the Word document is submitted to ISO).

**Comment by davaya** [comment]
@goneall
#1 Yes.
* I originally preferred a different Markdown format based on @sbarnum's table suggestion instead of nested lists, but I've changed my mind.  The "heading - level1 bullets - level2 bullets" format is the simplest possible format to parse and the simplest to standardize as the authoritative source format.
* The big advantage is that, as Ted Danson says, the Markdown format "has everything you need and nothing you don't".  Nobody needs to know a thing about OWL, SHACL, XML, XSD, JSON, or JSON Schema in order to read the markdown files, and barely more than nothing in order to write them.
* The biggest advantage is that the individual files are tiny and modular.  That makes them both easy to understand and easy to track under configuration management in GitHub.  The individual Lego blocks are composed into the model through subclassing, but I don't think it's possible to overstate the benefit of composing the model from those standardized little blocks.  @zvr: The ability to generate both ontologies and data schemas from the same set of blocks attests to the benefits of minimalism.

#2 For downstream processors, the "modelTypes.json" file looks like JSON but it really represents the Python `model` variable that contains all of the Lego blocks.  Instead of JSON it could have been the output of `print(model)` and the data would look the same except strings would be in single quotes. So I would separate part 2 into 3 parts:
* There should be a standard format for downstream processors that contains just the raw Lego blocks.  JSON is a universal format for representing application variables, but any format that preserves the individual blocks would do.
* There may also be a standard format for downstream processors that translates the Legos into SHACL/XSD (without OWL) - that's what the Markdown files represent.
* There may also be a standard format for downstream processors that translates the Legos into OWL.

I don't object to a standard OWL format, but I wouldn't use it because CBOR and Tag and Spreadsheets are not OWL serializations like RDF, JSON-LD and Turtle. My goal is to create CBOR payloads that transmit Element values without knowing or caring what graph the Elements represent.

**Comment by maxhbr** [comment]
(please do not reference random issues by using `#1` and `#2`.)

**Comment by davaya** [comment]
@zvr:
> Since we did not want to impose that everyone should learn these, the ad-hoc syntax was created as a convenience.

And yet that instinct toward simplicity and usability turned out to be the biggest advantage of the syntax.

The fact that the **model** as you see it can be created from the .md syntax proves that that syntax is *necessary and sufficient* to define the **model**.  The identical .md syntax defines the XSD/SHACL content that I care about without having to learn anything about OWL (i.e., it perfectly achieved your goal for creating it), and the fact that it is sufficient for your model means that the SHACL/XSD model automatically remains in sync.

**Comment by davaya** [comment]
@goneall

When you look at the your OWL/SHACL output, does it generate the identical content for a Relationship element as the example above (18 properties from 'from` to 'imports`, not 17 or 19), each of the properties of the correct type?

If using an OWL/SHACL/XSD model produces the identical payloads as a SHACL/.XSD-only model generated from the same Markdown files, then SPDX applications developed without any knowledge of OWL/RDF are completely interoperable with applications developed with the OWL model.

> * It allows us to change the input markdown format to meet the needs of human writability (one of the original goals)  
> * It allows downstream processors to depend on a well documented standard which will not change frequently  
> * Downstream processors can take advantage of existing parsers and validation libraries

The approach would be to develop the OWL/RDF model parser as you are already doing,  and document well the (trivial) markdown syntax that it uses.  There are no changes to your approach, and documenting the markdown syntax is no worse than leaving it undocumented.

The simple parser has no trouble with generating duplicate properties - the model markdown files don't call for them and the simple Relationship example parsed from them doesn't create them, so that's an example of how multiple implementations of the same spec uncovers things to be fixed, in the spec or in the software.

**Comment by goneall** [comment]
@davaya

> When you look at the your OWL/SHACL output, does it generate the identical content for a Relationship element as the example above (18 properties from from to imports, not 17 or 19), each of the properties of the correct type, and supporting simple types like DateTime so every time property doesn't have to have its own copy of the same string matching pattern?

Yes and no.  Yes, it is correct.  No it doesn't refer to the 18 properties since it takes into account the class hierarchy, so it only has the direct properties of the concrete class.  If you take into account the superclasses, I believe it is correct.

> The assertion that https://github.com/spdx/spdx-3-model/pull/377#issuecomment-1604009936, presumably because the spec-parser code for OWL/RDF doesn't handle them properly, is an annoyance.

As I've stated previously, we have the luxury of changing the definition and interpretation of the markdown files to address these issues.  My concern with having multiple interpretations of the markdown files is that this luxury goes away.

The reason in 2.0 I chose RDF OWL as the main representation of the model is that it was a superset of the data and ontology definitions.  In other words, you can go from RDF/OWL to XSD or JSON schema without loss, but not the other direction.

**Comment by davaya** [comment]
But you can go from markdown to RDF/OWL without loss because that's what the spec-parser already does.  And it's also possible to go from markdown to XSD and JSON schema.  And it's also possible to go from markdown to documentation web pages without loss.

There is always the luxury of adding things to the markdown.  And there won't be a need to change the things that are already there because they are based on SHACL and XSD.  Adding the "String" root type to Classes required adding the `schema` statement to the markdown Format section, in addition to the `pattern` statement that was already there.  There might be a few more things in SHACL or XSD that we want to pull in - they can always be added without breaking anything that is already there.

Document the definition and interpretation of the markdown files and feel free to update them as necessary to generate the RDF/OWL.  Feel free to add more sections like `Summary` and `Description` if necessary to create additional output products - the OWL and XSD and JSON Schema generators will ignore them.  Just document what we have and don't let that constrain your freedom to update.  I'm pretty sure there will never be a need for "markdown 2.0" because all the freedom you need won't require changing what's already there.

**Comment by davaya** [comment]
I've started working on a short paper called "The Class Model" that intends to define the SHACL/XSD and class hierarchy parts of the model, i.e. 90% of what the markdown represents.  (The other 10% being documentation - summary and description.)  I intend to include more XSD content (like the fact that simple datatypes can be restricted in three ways) that we don't use in the current markdown but could be used in the future.

The class model spec can of course be updated if needed, but XSD is pretty mature so hopefully the model spec will be stable and inclusive enough to use as a reference.  Any OWL/RDF material would be a generated supplement to the class model, not a competing or diverging source of truth.

**Comment by goneall** [comment]
We've made a decision to use the SHACL file for the "source of truth" and add annotations as needed - perhaps as a separate file.

Closing this issue as resolved.
